### PR TITLE
ENT-4288 Teach Tom to update dependency table

### PR DESCRIPTION
**Pull request by Lex-2008**
This updates tables in this file (only *master* branch):
https://github.com/Lex-2008/buildscripts/blob/master/deps-packaging/README.md

Command to run:
```
<@cf-bottom> depstable: 3.10.x,3.12.x,master
```

Example PR generated by the above command: https://github.com/cfengine/buildscripts/pull/560

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a PR! Maybe @olehermanse can review this?

File: tom/dependencies.py [hunk 1]
```diff
@@ -129,7 +135,10 @@ def maybe_replace(self, string, match, old, new):
 
     def extract_version_from_filename(self, dep, filename):
         if dep == 'openssl':
-            version = re.search('-([0-9a-z.]*).tar', filename).group(1)
+            try:
+                version = re.search('-([0-9a-z.]*).tar', filename).group(1)
+            except:
+                version = re.search('_([0-9a-z_]*).tar', filename).group(1)
```
Comment (vpodzime) [inline tom/dependencies.py]: Please do just `re.search`, check the result, then try the other one if it is `None` and then do `group(1)` on the match.

File: tom/dependencies.py [hunk 2]
```diff
@@ -235,6 +258,93 @@ def update_single_dep(self, dep):
...
+        upstream_name = 'cfengine'
+        local_path = "../" + repo_name
+        self.buildscripts = GitRepo(local_path, repo_name, upstream_name, self.username, 'master')
+
+        branches = branches.split(',')
+
+        # fetch versions from all branches
+        deps_table = {}
+        # deps_table is a 2d dict: deps_table[dep][branch]=version
+        branch_cloumn_widths = {}
```
Comment (vpodzime) [inline tom/dependencies.py]: `column`

File: tom/dependencies.py [hunk 3]
```diff
@@ -235,6 +258,93 @@ def update_single_dep(self, dep):
...
+                if dep not in deps_table:
+                    log.warn("unknown dependency in README: [%s] line [%s], will be EMPTY", dep, line)
+                    deps_table[dep] = {}
+                if has_notes:
+                    note = re.search('\| [^|]* \|$', line)
+                    if not note:
+                        log.warn("didn't find note in line [%s]", line)
+                        note = '|  |'
+                    note = note.group(0) # group(0) is full matched string
+                branches_text = ' | '.join([(deps_table[dep][branch] if branch in deps_table[dep] else '-').ljust(branch_cloumn_widths[branch]) for branch in branches])
```
Comment (Lex-2008) [inline tom/dependencies.py]: Any thoughts on how to wrap this line nicely?

File: tom/dependencies.py [hunk 4]
```diff
@@ -235,6 +266,93 @@ def update_single_dep(self, dep):
...
+                if dep not in deps_table:
+                    log.warn("unknown dependency in README: [%s] line [%s], will be EMPTY", dep, line)
+                    deps_table[dep] = {}
+                if has_notes:
+                    note = re.search('\| [^|]* \|$', line)
+                    if not note:
+                        log.warn("didn't find note in line [%s]", line)
+                        note = '|  |'
+                    note = note.group(0) # group(0) is full matched string
+                branches_text = ' | '.join([(deps_table[dep][branch] if branch in deps_table[dep] else '-').ljust(branch_column_widths[branch]) for branch in branches])
```
Comment (vpodzime) [inline tom/dependencies.py]: You can drop the `[ ]` inside the `join()` call. And I would wrap it before the `for`. And maybe use [`defaultdict`](https://docs.python.org/3.7/library/collections.html#collections.defaultdict) for `deps_table` like this:

```
deps_table = defaultdict(lambda: "-")
```
and then you can drop the `if-else`.

File: tom/dependencies.py [hunk 5]
```diff
@@ -235,6 +266,93 @@ def update_single_dep(self, dep):
...
+        self.readme_file_path = 'deps-packaging/README.md'
+        readme_file = self.buildscripts.get_file(self.readme_file_path)
+        readme_lines = readme_file.split('\n')
+        has_notes = False # flag to say that we're in a table that has "Notes" column
+        for i, line in enumerate(readme_lines):
+            if not line.startswith('| '):
+                continue
+            if line.startswith('| CFEngine version '):
+                has_notes = 'Notes' in line
+                branches_text = ' | '.join([branch.ljust(branch_column_widths[branch]) for branch in branches])
```
Comment (vpodzime) [inline tom/dependencies.py]: Extra `[]`s inside the `join()` call.

**Pull request merged**